src/packet_analysis/Analyzer.h

```diff
@@ -0,0 +1,148 @@
+// See the file "COPYING" in the main distribution directory for copyright.
+#pragma once
+
+#include "Manager.h"
+#include "Tag.h"
+#include <iosource/Packet.h>
+
+namespace zeek::packet_analysis {
+
+/**
+ * Result of packet analysis.
+ */
+ //TODO: Replace with bool?
+enum class AnalyzerResult {
+	Failed,   // Analysis failed
+	Terminate // Analysis succeeded and there is no further analysis to do
+};
+
+using AnalysisResultTuple = std::tuple<AnalyzerResult, uint32_t>;
+
+class Analyzer {
+public:
+	/**
+	 * Constructor.
+	 *
+	 * @param name The name for the type of analyzer. The name must match
+	 * the one the corresponding Component registers.
+	 */
+	explicit Analyzer(std::string name);
+
+	/**
+	 * Constructor.
+	 *
+	 * @param tag The tag for the type of analyzer. The tag must map to
+	 * the name the corresponding Component registers.
+	 */
+	explicit Analyzer(const Tag& tag);
+
+	/**
+	 * Destructor.
+	 */
+	virtual ~Analyzer() = default;
+
+	/**
+	 * Initialize the analyzer. This method is called after the configuration
+	 * was read. Derived classes can override this method to implement custom
+	 * initialization.
+	 */
+	virtual void Initialize() { };
+
+	/**
+	 * Returns the tag associated with the analyzer's type.
+	 */
+	const Tag GetAnalyzerTag() const;
+
+	/**
+	 * Returns a textual description of the analyzer's type. This is
+	 * what's passed to the constructor and usually corresponds to the
+	 * protocol name, e.g., "ARP".
+	 */
+	const char* GetAnalyzerName() const;
+
+	/**
+	 * Returns true if this analyzer's type matches the name passes in.
+	 * This is shortcut for comparing GetAnalyzerName() with the given
+	 * name.
+	 *
+	 * @param name The name to check.
+	 */
+	bool IsAnalyzer(const char* name);
+
+	/**
+	 * Registers an analyzer to be dispatched for the given identifier.
+	 *
+	 * @param identifier The identifier an analyzer should be called for.
+	 * @param analyzer The analyzer that should be called.
+	 */
+	void RegisterAnalyzerMapping(uint32_t identifier, AnalyzerPtr analyzer);
+
+	/**
+	 * Registers a default analyzer.
+	 *
+	 * @param default_analyzer The analyzer to use as default.
+	 */
+	void RegisterDefaultAnalyzer(AnalyzerPtr default_analyzer);
+
+	/**
+	 * Analyzes the given packet. The data reference points to the part of the
+	 * raw packet to be analyzed. If the analyzed protocol encapsulates another
+	 * protocol, the data reference should be updated to point to that payload.
+	 *
+	 * @param packet The packet to analyze.
+	 * @param data Reference to the payload pointer into the raw packet.
+	 *
+	 * @return A tuple of analysis result and identifier. The result indicates
+	 * how to proceed. If analysis can continue, the identifier determines the
+	 * encapsulated protocol.
+	 */
+	virtual AnalyzerResult Analyze(Packet* packet, const uint8_t*& data) = 0;
```

Comment:
I realize I'm getting nit-picky here, but some API suggestions because I think there's value in consistency with the other analyzer APIs:

-  What would you think about  passing the number of. bytes  available in `data` as a 3rd parameter? That's what the other analyzer APIs do, and it would make the out-of-bounds check in the methods more straight-forward: (`if ( len >= 8 )` instead of currently `if ( data + 8 >= packet->GetEndOfData() )`?

- I'd also change the order of the arguments to pass `data` first, because that's the main thing the method will be working on, whereas the `Packet` retains additional meta data.

- The other analyzer APIs use `u_char*` instead of `uint8_t*`.

- The standard Analyzer API names this `DeliverPacket()`

- Right now we have two cases where `AnalyzerResult::Terminate` is returned: (1) when packet analysis has finished successfully and wants to move on to the next layer; and (2) when an error occurs. The distinction is made by whether an call to Weird() has been made (correct?), which is pretty indirect (and also prevents using Weird() for non-fatal errors). How about we use the return value instead to pass back the new `data` (if we want to proceed), or nullptr if we got an error? That would then also avoid having to pass `data` by reference, and would be overall lead to a more natural data flow I believe. 


Taking this all together, we'd arrive at this:

```
virtual  const u_char* DeliverPacket(int len, const u_char* data, Packet* packet);
```

Would that make sense?

Reply:
> What would you think about passing the number of. bytes available in `data` as a 3rd parameter?

I don't have too strong feelings here. I just decided against this approach, because an analyzer that screws up calculating the remaining length would probably cause problems in one of the following analyzers. The root cause might get harder to find in this case.

> I'd also change the order of the arguments...

Ack.

> The other analyzer APIs use `u_char*` instead of `uint8_t*`.

Using `uint8_t*` seems to be the preferred variant nowadays (see https://stackoverflow.com/questions/1725855/uint8-t-vs-unsigned-char). Maybe we can think of adapting the other APIs ;)

> The standard Analyzer API names this `DeliverPacket()`

I thought about picking up that naming but to be honest I don't like it. In my mental model a method call to an object tells the object what to do (e.g. `rectangle.Draw()` - the rectangle takes care of drawing itself). So calling the `DeliverPacket()` method on an analyzer makes me wonder to whom the analyzer delivers the packet. To the next analyzer? But then, what's the difference between `DeliverPacket()` and `ForwardPacket()`? The perspective is changing: *We* deliver something to the analyzer but *the analyzer* forwards data to others. I think something like `ProcessPacket()` or `AnalyzePacket()` would be more intuitive. However, I am fine with renaming to be more consistent with the existing APIs.

> Right now we have two cases where `AnalyzerResult::Terminate` is returned: (1) when packet analysis has finished successfully and wants to move on to the next layer; and (2) when an error occurs.

Actually `AnalyzerResult::Terminate` is returned if things went well. In case of an error `AnalyzerResult::Failed` is returned. But you are 100% right with respect to the `Weird()` calls. Setting `l2_valid` controls whether we continue processing in the traditional stack. That should be coupled to the return value and not to a call to `Weird()`.
Overall there are three cases:
1. The analysis was successful and we want to pass the payload up into the traditional zeek stack.
2. There was an error, we want to signal that error and we don't want to pass the packet up.
3. The analysis was successful but we don't need pass the payload up into the traditional zeek stack. This is useful for protocols like ARP that are completely handled on packet analysis level.

The question is whether we need to differentiate outside of the affected analyzer between 2 and 3. If not, we could just return a `nullptr` in both cases. I can imagine one situation in which that wouldn't work: Let's say we have a packet analyzer and are not sure about what comes next. We want to try two possible options, one after the other (kind of poor man's DPD). If we cannot distinguish whether the first analyzer failed or terminated gracefully, we cannot decide whether to forward to the second one. I admit, right now this scenario is artificial but still maybe worth a thought.

Reply:
> I don't have too strong feelings here. I just decided against this approach, because an analyzer that screws up calculating the remaining length would probably cause problems in one of the following analyzers. The root cause might get harder to find in this case.

We could still do the calculation for the remaining length inside `AnalyzeInnerPacket` before passing control on to the next analyzer.


> Using `uint8_t*` seems to be the preferred variant nowadays 

True. Trade-off between consistency and modernization. :-)

> I thought about picking up that naming but to be honest I don't like it. 

Fair enough. Tim made a similar argument.  I would still prefer `Deliver` for the consistency across our APIs, but am fine being overruled here if you guys think it's bad. Same for the other notes on consistency: would prefer, but won't insist. :)

> 1. The analysis was successful and we want to pass the payload up into the traditional zeek stack.

This case will probably go away if we will eventually have IP forward directly into the traditional stack.

> The question is whether we need to differentiate outside of the affected analyzer between 2 and 3.

Yeah., I can see making that distinction for the return value, doesn't hurt to keep that information. So I would say this:

1. We need way to signal the difference cases of returns. If we have 2, that's a boolean; if that's >2, that's a enum.
2. If we need to pass the modified `data` back (see question above), then we should do it through the return value as well, not through a pass-by-reference parameter.

If we can express (1) and (2) through a single pointer that's potentially null, let's do that. If not, and we need to pass data, then let's return a tuple `(enum, new-data)`.